### PR TITLE
Replace MySQL prepared statement union with std::variant

### PR DESCRIPTION
**Pull request by Lillecarl**
**Changes proposed:**
-  Replace union with std::variant

**Target branch(es):**

- [x] 3.3.5
- [ ] master

**Issues addressed:**
Using more modern C++ instead of old C-style

**Tests performed:**
Builds, logged ingame, saved character, created account (tested as these will run MySQLPreparedStatement::setString)

### CODE REVIEW

**Comment by jackpoz** [comment]
could you merge 3.3.5 into this PR to fix the build ?

**Comment by jackpoz** [comment]
Could you paste the command you used to spot this btw ?

**Comment by Lillecarl** [comment]
Nope, this is from the iso cpp guidelines
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Ru-naked
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es48-avoid-casts

**Comment by jackpoz** [comment]
it doesn't seem like you merge all 3.3.5 commits as it still complains about DoCast()

**Comment by Lillecarl** [comment]
@jackpoz Ah yes, I've told myself to learn the Git commandline properly instead of point n click guis. Certain things you miss/forget still (git fetch doesn't fetch all remotes in this case)

I also rebased the work off the last TC version now.

Cheers :)

File: src/server/database/Database/MySQLPreparedStatement.cpp [hunk 1]
```diff
@@ -214,48 +226,48 @@ std::string MySQLPreparedStatement::getQueryString() const
     std::string queryString(m_queryString);
 
     size_t pos = 0;
-    for (uint32 i = 0; i < m_stmt->statement_data.size(); i++)
+    for (PreparedStatementData data : m_stmt->statement_data)
```
Comment (Shauren) [inline src/server/database/Database/MySQLPreparedStatement.cpp]: copy

Comment (Lillecarl) [reply src/server/database/Database/MySQLPreparedStatement.cpp]: Good catch.

**Comment by jackpoz** [comment]
> (git fetch doesn't fetch all remotes in this case)

"git remote update" refreshes everything :)

**Comment by Lillecarl** [comment]
Thanks for the tip :)

**Comment by Lillecarl** [comment]
Is there anything else i can do?

File: src/server/database/Database/PreparedStatement.cpp [hunk 2]
```diff
@@ -33,54 +33,55 @@ void PreparedStatement::BindParameters(MySQLPreparedStatement* stmt)
     ASSERT(stmt);
     m_stmt = stmt;
 
-    uint8 i = 0;
-    for (; i < statement_data.size(); i++)
+    uint8 pos = 0;
+    for (PreparedStatementData data : statement_data)
```
Comment (jackpoz) [inline src/server/database/Database/PreparedStatement.cpp]: is this another case of "copy" ?

Comment (Lillecarl) [reply src/server/database/Database/PreparedStatement.cpp]: That should've been fixed already, but yes 😅

**Review by jackpoz — CHANGES_REQUESTED** [review]
Please fix the PreparedStatement by reference iterators

**Comment by jackpoz** [comment]
Thanks for the PR :)

**Pull request merged**